### PR TITLE
Map IntervalList format column four to feature name

### LINKED ISSUE #1152: Map IntervalList format column four to feature name or attributes?
We are not sure if the name column should be mapped to `Feature.name` or to `Feature.attributes`.

Note discussion here
https://github.com/bigdatagenomics/adam/pull/1104/files#r77733324
https://github.com/bigdatagenomics/adam/pull/1104/files#r74367312

Ping @ryan-williams to comment.

### PR DESCRIPTION
**Pull request by heuermh**
Fixes #1152, #1168

### CODE REVIEW

File: adam-core/src/main/scala/org/bdgenomics/adam/rdd/features/FeatureParser.scala [hunk 1]
```diff
@@ -67,15 +54,28 @@ private[rdd] object GTFParser {
...
+    }
 
     val fields = line.split("\t")
-    // skip empty or invalid lines
+
+    // check for invalid lines
     if (fields.length < 8 || fields.length > 9) {
-      log.warn("Empty or invalid GTF/GFF2 line: {}", line)
-      return Seq()
+      if (stringency == ValidationStringency.STRICT) {
```
Comment (fnothaft) [inline adam-core/src/main/scala/org/bdgenomics/adam/rdd/features/FeatureParser.scala]: As an aside, I feel like we've got enough of this patten (print error message if STRICT, log if LENIENT, and return None if not STRICT) around where we might as well wrap it up into a function and factor it out.

File: adam-core/src/main/scala/org/bdgenomics/adam/rdd/features/FeatureParser.scala [hunk 2]
```diff
@@ -327,35 +338,44 @@ private[rdd] class BEDParser extends FeatureParser {
...
-    val fb = Feature.newBuilder()
-    fb.setContigName(fields(0))
 
-    // Peak files are 0-based space-coordinates, so conversion to
-    // our coordinate space should mean that the values are unchanged.
-    fb.setStart(fields(1).toLong)
-    fb.setEnd(fields(2).toLong)
+    val f = Feature.newBuilder()
+      .setContigName(fields(0))
+      .setStart(fields(1).toLong) // NarrowPeak ranges are 0-based
```
Comment (fnothaft) [inline adam-core/src/main/scala/org/bdgenomics/adam/rdd/features/FeatureParser.scala]: We should try/catch around the `fields(x).toLong`/`fields(x).toDouble` calls, both here and below and above and everywhere. We could factor it out like so:

```
def checkAndSet[T](field: String,
  convFn: String => T,
  setFn: T => Unit,
  stringency: ValidationStringency) {
  try {
    val t = convFn(field)
    setFn(t)
  } catch {
    case e: Throwable => {
      if (stringency == ValidationStringency.LENIENT) {
        log.warn("Failed to convert %s.".format(field))
      } else if (stringency == ValidationStringency.STRICT) {
        throw new IllegalArgumentException("Setting field from %s failed with %s.".format(
          field,
          e.msg))
      }
    }
  }
}
```

Perhaps we factor this out into a trait:

```
sealed trait FeatureConverterCanBeValidated {
  val stringency: ValidationStringency
  
  def validateOrNone(...): Option[Feature]

  def checkAndSet[T](field: String,
    convFn: String => T,
    setFn: T => Unit)
}
```

All of the classes in here could be turned into case classes that extend that trait. What do you think? Should you proceed, the trait name I proposed is terrible, so please change it.

**Comment by fnothaft** [comment]
Generally LGTM! Thanks for taking this on @heuermh. I've dropped a variety of line notes inline.

File: adam-core/src/main/scala/org/bdgenomics/adam/rdd/features/FeatureParser.scala (hunk 1 above)
Comment (heuermh) [inline adam-core/src/main/scala/org/bdgenomics/adam/rdd/features/FeatureParser.scala]: Agreed

File: adam-core/src/main/scala/org/bdgenomics/adam/rdd/features/FeatureParser.scala (hunk 2 above)
Comment (heuermh) [inline adam-core/src/main/scala/org/bdgenomics/adam/rdd/features/FeatureParser.scala]: Will check to see what htsjdk does with NumberFormatExceptions and validation stringency.

File: adam-core/src/main/scala/org/bdgenomics/adam/rdd/features/FeatureParser.scala (hunk 1 above)
Comment (fnothaft) [inline adam-core/src/main/scala/org/bdgenomics/adam/rdd/features/FeatureParser.scala]: How so?

File: adam-core/src/main/scala/org/bdgenomics/adam/rdd/features/FeatureParser.scala (hunk 2 above)
Comment (heuermh) [inline adam-core/src/main/scala/org/bdgenomics/adam/rdd/features/FeatureParser.scala]: This seems to fiddly to me, so I've punted for now

**Comment by heuermh** [comment]
Addressed some review comments and rebased.

**Comment by fnothaft** [comment]
LGTM! I will leave this open until Friday morning to wait for any further review comments.

**Comment by fnothaft** [comment]
Thanks @heuermh! Merged as 53b8f48abe0b0536c39e155fb32de91a98c78876 and dff43effd9f52f8cdcec8952a0a27c5c258afd45.

**Comment by heuermh** [comment]
Thanks!

**Pull request closed without merging**